peer_test.go

```diff
@@ -217,3 +219,268 @@ func TestPeerSelectionPreferIncoming(t *testing.T) {
+	numCalls int
+	// numAffinity is the number of affinity nodes.
+	numAffinity int
+	// numConcurrent is the number of concurrent gocoroutine.
+	numConcurrent int
+
+	servers  []*Channel
+	affinity []*Channel
+	client   *Channel
+}
+
+// setupServers will create numPeer servers, and register handlers on them.
+func (pt *peerSelectionTest) setupServers(t *testing.T) {
+	pt.servers = make([]*Channel, pt.numPeers)
+
+	// Set up numPeers servers.
+	for i := 0; i < pt.numPeers; i++ {
+		pt.servers[i], _ = pt.NewService(t, "hyperbahn")
+		pt.servers[i].Register(raw.Wrap(newTestHandler(pt.t)), "echo")
+	}
+	fmt.Println("all server nodes setup")
+}
+
+func (pt *peerSelectionTest) setupAffinity(t *testing.T) {
+	var wg sync.WaitGroup
+	pt.affinity = make([]*Channel, pt.numAffinity)
+	for i := range pt.affinity {
+		pt.affinity[i] = pt.servers[i]
+	}
+
+	for _, server := range pt.servers {
+		pt.client.Peers().Add(server.PeerInfo().HostPort)
+	}
+	fmt.Println("added all hyperbahn nodes")
+
+	wg.Add(pt.numAffinity)
+	// Connect from the affinity nodes to the service.
+	for _, affinity := range pt.affinity {
+		affinity.Peers().Add(pt.client.PeerInfo().HostPort)
+		go func(affinity *Channel) {
+			pt.makeCall(affinity.GetSubChannel(pt.client.PeerInfo().ServiceName))
+			wg.Done()
+		}(affinity)
+	}
+	wg.Wait()
+	fmt.Println("affinity nodes connected")
+}
+
+func (pt *peerSelectionTest) setupClient(t *testing.T) {
+	pt.client, _ = pt.NewService(t, "client")
+	pt.client.Register(raw.Wrap(newTestHandler(t)), "echo")
+	for _, server := range pt.servers {
+		pt.client.Peers().Add(server.PeerInfo().HostPort)
+	}
+	time.Sleep(time.Millisecond * 100)
+}
+
+func (pt *peerSelectionTest) makeCall(sc *SubChannel) {
+	ctx, cancel := NewContext(time.Second)
+	defer cancel()
+	_, _, _, err := raw.CallSC(ctx, sc, "echo", nil, nil)
+	require.NoError(pt.t, err, "raw.Call failed")
+}
+
+func (pt *peerSelectionTest) runStress() {
+	var wg sync.WaitGroup
+
+	// helper that will make a request every n ticks.
+	reqEveryNTicks := func(n int, sc *SubChannel, clock <-chan struct{}) {
+		defer wg.Done()
+
+		for {
+			for i := 0; i < n; i++ {
+				_, ok := <-clock
+				if !ok {
+					return
+				}
+			}
+			pt.makeCall(sc)
+		}
+	}
+
+	numClock := pt.numConcurrent + pt.numAffinity
+	clocks := make([]chan struct{}, numClock)
+	for i := 0; i < numClock; i++ {
+		clocks[i] = make(chan struct{})
+	}
+	wg.Add(numClock)
+
+	// server outbound request
+	sc := pt.client.GetSubChannel("hyperbahn")
+	for i := 0; i < pt.numConcurrent; i++ {
+		go reqEveryNTicks(1, sc, clocks[i])
+	}
+	// affinity incoming requests
+	for i := 0; i < pt.numAffinity; i++ {
+		go reqEveryNTicks(1, pt.affinity[i].GetSubChannel(pt.client.PeerInfo().ServiceName), clocks[i+pt.numConcurrent])
+	}
+
+	tickAllClocks := func() {
+		for i := 0; i < numClock; i++ {
+			clocks[i] <- struct{}{}
+		}
+	}
+
+	const numTicks = 10000
+	for i := 0; i < numTicks; i++ {
+		if i%(numTicks/10) == 0 {
```

Comment:
don't print this so often, probably want to make this more like `1000` or maybe `500`